keps/sig-api-machinery/3488-cel-admission-control/README.md

```diff
@@ -0,0 +1,2862 @@
+      operations:  ["CREATE", "UPDATE"]
+      resources:   ["deployments"]
+```
+
+Since this policy is constrainted to create/update of deployments. Policy
+bindings don't need to repeat this constraint. This would be sufficient:
+
+```yaml
+apiVersion: admissionregistration.k8s.io/v1alpha1
+kind: PolicyBinding
+...
+spec:
+  match:
+    namespaceSelectors:
+    - key: environment
+      operator: Exists
+ ...
+```
+
+We can enforce this by requiring:
+
+- Policy definitions match rules are validated to match one or more apiGroup +
+  resource + operation.
+- Policy bindings are not required to match apiGroup + resource +
+  operation since the policy definition is already required to do this.
+
+This encourages policy definition authors to consider the assumptions that the
+CEL expressions make. If the expressions unconditionally access `object` without
+a `has(object)` check, the expression will only ever work on `CREATE` and
+`UPDATE` and would fail at runtime on a `DELETE`. It is quite difficult to write
+CEL expression that handle all admission requests well, so we want to guide
+policy authors toward matching only the requests that they intend to support.
+The Kubernetes admission chain also scales better, and is more resiliant, when
+matching is precise and the validation expressions don't need to do any
+post-matching checks that could have been handled by matching.
+
+Note that if a policy binding is to be applied as broadly as possible (i.e.
+everywhere allowed by the policy definition) it must do so by using a wildcard
+match rule. An `effect: MatchAll` (only permitted if no other match criteria is
+present) could be offered to make this easy, or an existing match criteria that
+supports wildcards could be used.
+
+Match Policy:
+
+`MatchPolicy` will work the same as for admission webhooks. It will default to
+`Equivalent` but may be set to `Exact`. See "Use Case: Multiple policy
+definitions for different versions of CRD" for an explanation of why we need
+`MatchPolicy`.
+
+xref:
+
+- https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/#matching-requests-rules
+- https://kubernetes.io/docs/tasks/debug/debug-cluster/audit/
+- https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.25/#resourcepolicyrule-v1beta1-flowcontrol-apiserver-k8s-io
+- https://github.com/open-policy-agent/gatekeeper/blob/a1add93b0beb5c48eb92a6a2eb5ee7d21551a1b6/pkg/mutation/match/match.go#L22-L56
+
+#### Reporting violations to Clients
+
+This section focuses on how information is reported back to clients in
+when validations fail.
+
+Goals:
+
+- Feature parity with `AdmissionReview`
+  - Support allow/deny result, warnings, audit annotations
+  - Support for reasons/codes
+- Ability to format message strings
+
+High level proposal:
+
+- Each policy may set a `denyReason` and/or `denyCode`
+- Each validation may define a message:
+  - `message` - may contain `{1}` format args that are supplied by
```

Comment:
Any reason why the message can't be a CEL expression itself?

Reply:
We looked at this as an option, see "alternatives considered" and my above, I'm happy to reconsider this decision as I don't feel super strong about the alternatives. Ideally we come up with a plan that works for CRD validation as well.

Reply:
I've added a UNRESOLVED to this section so we don't approve this design until we decide on an approach here.

Reply:
Switched to `message` and `messageExpression`